### PR TITLE
Stocks database update unit tests PR-28

### PR DESCRIPTION
**Pull request by Malak-abuthaa**
1. some enhancement on scheduler.py , also add function to stop all jobs in order to control it in the test
2. add test_scheduler.py , that test number of thread that lunch with the scheduler 
3. add test_jobs.py , to to check if value last_modified has been changed when the stock_api_update called 
4. new field has been added to the stock model 
5. __init__ + app.py , has been been modified to meet the scheduler enhancement

### CODE REVIEW

File: myapp/models.py [hunk 1]
```diff
@@ -3,33 +3,36 @@
...
-		profile.save()
+    symbol = models.CharField(max_length=12, primary_key=True)
+    name = models.CharField(max_length=64)
+    top_rank = models.IntegerField(null=True)
+    price = models.FloatField()
+    change = models.FloatField(null=True)
+    change_percent = models.FloatField(null=True)
+    market_cap = models.FloatField(null=True)
+    primary_exchange = models.CharField(null=True, max_length=32)
+    last_modified = models.DateTimeField(auto_now=True)
```
Comment (kamalq97) [inline myapp/models.py:17]: Maybe we can also show the 'last modified' field on the relevant html template as well 👍

File: myapp/tests/test_jobs.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class test_api(TestCase):
+
+    def test_api_update(self):
+        stock_api_update()
+        value_1 = Stock.objects.filter(top_rank=1)[:1][0].last_modified
+        stock_api_update()
+        value_2 = Stock.objects.filter(top_rank=1)[:1][0].last_modified
+        self.assertNotEqual(value_1, value_2)
```
Comment (kamalq97) [inline myapp/tests/test_jobs.py:13]: Yeah, this is what I was talking about in the meeting yesterday, nice!  Maybe you could use `assertAlmostEqual()` to check if the time difference is approximately 10 seconds as opposed to just checking if the times are not the same?

**Review by kamalq97 — APPROVED** [review]
Seems good overall. Consider changing the assert statement in `test_api_update` as I suggested in the comment below.

_!!!! **IMPORTANT**: It seems like some of the changes you have made are resulting in migration errors, **please try to sort these out before merging**!_

File: myapp/migrations/0002_auto_20190920_1142.py [hunk 3]
```diff
@@ -1,18 +0,0 @@
-# Generated by Django 2.2.4 on 2019-09-20 08:42
```
Comment (murradkh) [inline myapp/migrations/0002_auto_20190920_1142.py]: I'm not sure why migration has been deleted, in general, that's should not happen! this is causing errors for me and kamal when running the "migrate" command due to prior data existed in our database.
it's fine to keep like this for this PR. 
@kamaql97 delete the data in DB file then run the migration should be run without error.

File: myapp/tests/test_jobs.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from django.test import TestCase
+from ..scheduler.stock_api_update import stock_api_update
+from ..models import Stock
+
+
+class test_api(TestCase):
```
Comment (murradkh) [inline myapp/tests/test_jobs.py:6]: camel-case needs to be like `SchedulerJobsTestCase`
actually, I don't think this file is needed, it's relevant to "scheduler", so added inside `test_scheduler.py` please

File: myapp/tests/test_jobs.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+    def test_api_update(self):
+        stock_api_update()
+        value_1 = Stock.objects.filter(top_rank=1)[:1][0].last_modified
```
Comment (murradkh) [inline myapp/tests/test_jobs.py:10]: nice!
I'm not sure why you slicing the array `[:1]` and then `[0],` why not just `[0]`

File: myapp/tests/test_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from django.test import TestCase
+from ..scheduler import scheduler
+import threading
+
+
+
+class test_scheduler(TestCase):
+
+    def test_scheduler_thread(self):
+        one = threading.active_count()
```
Comment (murradkh) [inline myapp/tests/test_scheduler.py:11]: nice

File: myapp/tests/test_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+        job = scheduler()
+        job.start()
+        self.assertNotEqual(one, threading.active_count())
```
Comment (murradkh) [inline myapp/tests/test_scheduler.py:14]: be more specific and concrete please, how many threads you are going to create, and then use `assertEqual`

File: myapp/tests/test_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+        job.stop()
```
Comment (murradkh) [inline myapp/tests/test_scheduler.py]: this is not Working! (doesn't stop the threads)
please make comprehensive testing on your code before creating the PR!

**Review by murradkh — CHANGES_REQUESTED** [review]

File: myapp/tests/test_scheduler.py (hunk 8 above)
Comment (Malak-abuthaa) [reply myapp/tests/test_scheduler.py]: i will check it again

File: myapp/tests/test_scheduler.py (hunk 7 above)
Comment (Malak-abuthaa) [reply myapp/tests/test_scheduler.py:14]: three threads should be created , one for the main test , other two for scheduler , i can put three , but in the future make sure to increase that number

File: myapp/tests/test_jobs.py (hunk 2 above)
Comment (Malak-abuthaa) [reply myapp/tests/test_jobs.py:13]: in this section i'm testing the API not the job , so the difference is so little second or less

File: myapp/tests/test_scheduler.py (hunk 7 above)
Comment (murradkh) [reply myapp/tests/test_scheduler.py:14]: > three threads should be created , one for the main test , other two for scheduler , i can put three , but in the future make sure to increase that number

I agree with you on this point

**Review by murradkh — APPROVED** [review]
confirm no errors in migration /tests / server-running

**Pull request merged**